Move bearer token generation and outbound header writing

keeratsingh opened the pull request:
- Move bearer token generation and outbound header writing
- Abstract implementation details for BasicCallHeaderAuthenticator header authentication away in the AuthValidator interface.
- Remove getBearerToken from CallHeaderAuthenticator
- Create a generalized implementation of FlightClientMiddleware that taken in a handler,
- that knows how to handle incoming headers for the client.

Commit:
Move bearer token generation and outbound header writing

- Move bearer token generation and outbound header writing

Commit:
Refactored the authentication to validate basic and non basic headers

- Refactored the authentication to validate basic and non basic headers

Commit:
Abstract BasicCallHeaderAuthenticator header authentication.

- Abstract implementation details for BasicCallHeaderAuthenticator header authentication away in
in the AuthValidator interface.
- Remove getBearerToken from CallHeaderAuthenticator

Commit:
Create a generalized implementation of FlightClientMiddleware

- Create a generalized implementation of FlightClientMiddleware that taken in a handler,
that knows how to handle incoming headers for the client.

Commit:
Fix build failures

Fix build failures

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java:
```diff
@@ -0,0 +1,147 @@
+
+  @Override
+  public AuthResult validateIncomingHeaders(CallHeaders incomingHeaders) throws Exception {
+    AuthResult authResult = parseAndValidateBearerHeaders(incomingHeaders);
```
nit: final

keeratsingh replied:
Done.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java:
```diff
@@ -0,0 +1,147 @@
+    final String authDecoded = new String(Base64.getDecoder().decode(authEncoded), StandardCharsets.UTF_8);
+    final int colonPos = authDecoded.indexOf(':');
+    if (colonPos == -1) {
+      throw CallStatus.UNAUTHORIZED.toRuntimeException();
```
Should this be unauthenticated? Should this throw?

keeratsingh replied:
Yes, this should be UNAUTHENTICATED, updated code. I believe yes, this should throw so that authentication fails and the exception is caught in the authenticate method.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java:
```diff
@@ -0,0 +1,147 @@
+   * Interface that this validator delegates for determining if the credentials are valid.
+   */
+  interface CredentialValidator {
+    Optional<String> validate(String username, String password) throws Exception;
```
What does this return? Please add comments to these interface methods (and in AuthTokenManager)

keeratsingh replied:
Done, added comments.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java:
```diff
@@ -0,0 +1,147 @@
+
+    final String user = authDecoded.substring(0, colonPos);
+    final String password = authDecoded.substring(colonPos + 1);
+    credentialValidator.validate(user, password);
```
Why do we discard the return value?

keeratsingh replied:
Updated code to use the return value.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java:
```diff
@@ -0,0 +1,147 @@
+    return new AuthResult() {
+      @Override
+      public String getPeerIdentity() {
+        return user;
```
I think this is not correct. If we do this, then if the same user has multiple connections they share a single peer identity.

keeratsingh replied:
Updated code to return the peer Identity that is returned from after validating the credentials. So now it is up to the impl to decide what peer identity it returns incase of multiple connections for the same user.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java:
```diff
@@ -0,0 +1,147 @@
+    if (parsedValue != null) {
+      final Optional<String> peerIdentity = authTokenManager.validateToken(parsedValue);
+      if (!peerIdentity.isPresent()) {
+        throw CallStatus.UNAUTHORIZED.toRuntimeException();
```
Unauthenticated or unauthorized?

keeratsingh replied:
Updated to throw Unauthenticated.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BearerTokenAuthHandler.java:
```diff
@@ -26,9 +26,12 @@
  */
 abstract class BearerTokenAuthHandler implements CallHeaderAuthenticator {
+
+  String bearerToken;
```
There is a single CallHeaderAuthenticator that the server uses, so this logic does not work if you have multiple connections. Please add unit tests with multiple FlightClients connecting to the same server, both with differing credentials/identities and the same credentials.

keeratsingh replied:
Done, added tests for both BasicCallHeaderAuthenticator and GeneratedBearerTokenAuthHandler.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/ClientIncomingAuthHeaderMiddleware.java:
```diff
@@ -23,39 +23,48 @@
   public static class Factory implements FlightClientMiddleware.Factory {
-    private final AtomicReference<String> bearerToken = new AtomicReference<>();
+    private final ClientHeaderHandler headerHandler;
+    private final AtomicReference<CredentialCallOption> credentialCallOption = new AtomicReference<>();
```
I don't think this has to be an AtomicReference since this looks to be used single-threaded only.

keeratsingh replied:
Done, updated.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/GeneratedBearerTokenAuthHandler.java:
```diff
@@ -44,7 +44,7 @@ public boolean validateBearer(String bearerToken) {
 
   String registerBearer(AuthResult handshakeResult) {
-    final String bearerToken = handshakeResult.getBearerToken()
+    final String bearerToken = getBearerToken()
```
Why not create a specialized subclass of AuthResult that has getBearerToken() and pass that in to registerBearer? Where does registerBearer get called from now?

keeratsingh replied:
Done, updated authentication flow to call registerBearer.

Commit:
Update BearerTokenAuthHandler

- Updated BearerTokenAuthHandler
- Added tests for BearerTokenAuthHandler

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightClient.java:
```diff
@@ -192,13 +194,13 @@ public void authenticate(ClientAuthHandler handler, CallOption... options) {
-    middleware.add(tokenMiddleware);
+  public Optional<CredentialCallOption> authenticateBasicToken(String username, String password) {
+    final ClientIncomingAuthHeaderMiddleware.Factory clientAuthMiddleware =
+            new ClientIncomingAuthHeaderMiddleware.Factory(headerHandler);
```
Let's make this method use the bearer token header handler. This method is specifically for using the basic auth workflow that attempts to get a bearer token (see the javadoc above).

Someone that wants to use a different outgoing header can go setup a ClientIncomingAuthMiddleware themselves and call handshake themselves. Let's remove the ClientHeaderHandler from the builder.

keeratsingh replied:
Done.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java:
```diff
@@ -0,0 +1,194 @@
+      }
+
+      @Override
+      public CallHeaderAuthenticator.HeaderMetadata getHeaderMetadata() {
```
I don't see why we need this method. It looks like it caches information to use later in appendToOutgoingHeaders(), but you can just as easily use this information directly since getHeaderMetadata() and appendToOutgoingHeaders() are implemented at the same time.

I also see this being used in registerBearer which I'll comment on separately.

keeratsingh replied:
Removed the interface and the method.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java:
```diff
@@ -0,0 +1,194 @@
+
+  @Override
+  public AuthResult validateIncomingHeaders(CallHeaders incomingHeaders) throws Exception {
+    final AuthResult authResult = parseAndValidateBearerHeaders(incomingHeaders);
```
I don't think this is the right way to go. This incurs the overhead of checking for bearer tokens even if my server only supports basic auth w/o bearer tokens.

What you should do instead is have a CompositeAuthValidator that tries to use one credential validator first (in this case a bearer token validator) then tries to use the other one (in this case a basic validator).

You could potential chain more composites to support more authentication strategies.

Since the bearer token / basic auth case is common, you should add a static factory function to ServerAuthValidator to more easily create an appropriately configured CompositeAuthValidator.

keeratsingh replied:
Updated code to use a CompositeAuthValidator as suggested.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/ClientIncomingAuthHeaderMiddleware.java:
```diff
@@ -17,40 +17,52 @@
 import org.apache.arrow.flight.CallStatus;
 import org.apache.arrow.flight.FlightClientMiddleware;
+import org.apache.arrow.flight.grpc.CredentialCallOption;
+import org.slf4j.Logger;
```
I deleted these from the original class. You should rebase on the top of the flight-auth-redesign branch to make sure you are up to date.

keeratsingh replied:
Done, removed. I believe I added these back during a merge conflict in the rebase.

jduo commented on java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/GeneratedBearerTokenAuthHandler.java:
```diff
@@ -33,26 +36,29 @@
-    return bearerToken;
+  @Override
+  protected void registerBearer(AuthResult authResult) {
+    String bearerToken = authResult.getHeaderMetadata().getValue();
```
It seems like this is the only place where getHeaderMetadata() is used outside of the AuthResult's implementation. So really, this function (and corresponding interface) only exist for the purpose of servicing this method. Which we don't want, we are making the interface more complex so that we can handle one use case.

Instead, why don't we change this code to call authResult.appendToOutgoingHeaders() with a dummy CallHeaders object. Then just inspect the CallHeaders object to see if it has the bearer token header.

Note that your current code doesn't validate that getValue() is actually returning a bearer token.

keeratsingh replied:
Done, updated the code to call appendToOutgoingHeaders with dummy headers and parse the headers as suggested.

jduo commented on java/flight/flight-core/src/test/java/org/apache/arrow/flight/auth2/TestBasicAuth2.java:
```diff
@@ -49,117 +49,239 @@
+      }
+    };
+
+    serverWithBasicAuth = FlightTestUtil.getStartedServer((location) -> FlightServer
```
Let's not do this in setup() and just have a helper method for setting up test state that each test can invoke (ie parameters can be the auth type, number of clients, etc).

What you're doing is allocating everything for every test then cleaning everything up for every test. It seems like alot of overhead.

keeratsingh replied:
Updated tests to have better resource allocation.

jduo reviewed (changes requested):
See other comments.

Commit:
Address code review comments

- Modify BasicAuthValidator to have a factory that takes in a composite validator
- Remove HeaderMetadata
- Update registerBearer to use dummy headers instead of relying on HeaderMetadata.

Commit:
Update BasicAuthValidator to include CompositeCredentialValidator and clean up TestBasicAuth2

- Update BasicAuthValidator to include CompositeCredentialValidator.
- Clean up TestBasicAuth2 to make sure not all resources are allocated for all tests.

jduo committed:
Rework authenticator classes

jduo committed:
Obscure authentication failures from client
